### PR TITLE
Switch to JDK 18 as minumum supported version

### PR DESCRIPTION
**Pull request by deejay1**
Switches to 18 as base release having the minimum supported features. Adds matrix build for JDK versions up to 21.

### CODE REVIEW

**Comment by deejay1** [comment]
> Thanks for the PR.
> 
> I'm fine with adding Java 21 to the CI, but unless there's a reason to drop support for earlier versions, I'd prefer not to do so. Some users would like to even support Java 17, as mentioned in issue #28.
> 
> My suggestion would be to create a separate GitHub job that utilizes Java 21 for building and testing the library. We can set this up using a [matrix](https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs#using-a-matrix-strategy).

Switched to a matrix build and to 18 as the minimal version with all required features that are used now.

File: .github/workflows/publish.yml [hunk 1]
```diff
@@ -13,15 +13,15 @@ jobs:
...
 
       - uses: gradle/wrapper-validation-action@v1
 
-      - name: Set up JDK 20
+      - name: Set up JDK 21
         uses: actions/setup-java@v3
         with:
           distribution: temurin
-          java-version: 20
+          java-version: 21
```
Comment (piotrrzysko) [inline .github/workflows/publish.yml:24]: Is there a specific reason for using 21 here? Shouldn't we use the lowest supported version for publishing packages?

File: build.gradle [hunk 2]
```diff
@@ -31,7 +31,8 @@ repositories {
 
 java {
     toolchain {
-        languageVersion = JavaLanguageVersion.of(20)
+        languageVersion = JavaLanguageVersion.of(18)
+        vendor = JvmVendorSpec.ADOPTIUM
```
Comment (piotrrzysko) [inline build.gradle:35]: Why do we need the limitation to one vendor?

**Review by piotrrzysko — COMMENTED** [review]
In the  README there are two places where the required Java version is mentioned: 
* https://github.com/simdjson/simdjson-java/blob/main/README.md?plain=1#L50
* https://github.com/simdjson/simdjson-java/blob/main/README.md?plain=1#L90

If we want to start supporting Java 18, we need to update the README.

File: build.gradle (hunk 2 above)
Comment (deejay1) [reply build.gradle]: no

**Pull request merged**